File: src/lib/settler.ts
import { getCurrentUser } from "./identity";
import type { User } from "../db/queries/users";

function settlerHandles(): Set<string> {
  return new Set(
    (process.env.SETTLER_USERNAMES ?? "")
      .split(",")
      .map((entry) => entry.trim().toLowerCase().replace(/^@/, ""))
      .filter(Boolean),
  );
}

/** A settler is any logged-in user whose X username or twitter_id is configured. */
export function isSettler(user: User | null): boolean {
  if (!user) return false;
  const handles = settlerHandles();
  if (handles.size === 0) return false;
  return (
    (!!user.username && handles.has(user.username.toLowerCase())) ||
    (!!user.twitter_id && handles.has(user.twitter_id.toLowerCase()))
  );
}

export async function getCurrentSettler(): Promise<User | null> {
  const user = await getCurrentUser();
  return isSettler(user) ? user : null;
}

File: rails/app/models/settler.rb
# A settler is any logged-in user with a linked account whose provider handle
# or provider account id is listed in SETTLER_USERNAMES (comma-separated, "@"
# and case insensitive). Settlers can access the settlement admin.
module Settler
  module_function

  def handles
    (ENV["SETTLER_USERNAMES"] || "")
      .split(",")
      .map { |entry| entry.strip.downcase.delete_prefix("@") }
      .reject(&:empty?)
      .to_set
  end

  def settler?(user)
    return false unless user

    allowed = handles
    return false if allowed.empty?

    user.accounts.any? do |account|
      (account.username.present? && allowed.include?(account.username.downcase)) ||
        allowed.include?(account.provider_account_id.to_s.downcase)
    end
  end
end
